feature/experimental/gribi/ate_tests/dut_daemon_failure/dut_daemon_failure_test.go

```diff
@@ -0,0 +1,378 @@
+	t.Logf("Verify through gRIBI Get RPC that %s is present", ateDstNetCIDR)
+	getResponse, err := clientA.Fluent(t).Get().WithNetworkInstance(*deviations.DefaultNetworkInstance).WithAFT(fluent.IPv4).Send()
+	if err != nil {
+		t.Errorf("Cannot Get: %v", err)
+	}
+	entries := getResponse.GetEntry()
+	var found bool = false
+	for _, entry := range entries {
+		v := entry.Entry.(*grps.AFTEntry_Ipv4)
+		if prefix := v.Ipv4.GetPrefix(); prefix != "" {
+			if prefix == ateDstNetCIDR {
+				found = true
+				t.Logf("Found route to %s in gRIBI Get response", ateDstNetCIDR)
+				break
+			}
+		}
+	}
+	if !found {
+		t.Errorf("Route to %s NOT found in gRIBI Get response", ateDstNetCIDR)
+	}
+}
+
+// gNOIKillProcess kills a daemon on the DUT, given its name and pid.
+func gNOIKillProcess(ctx context.Context, t *testing.T, args *testArgs, pName string, pID uint32) {
+	gnoiClient := args.dut.RawAPIs().GNOI().Default(t)
+	killRequest := &gnps.KillProcessRequest{Name: pName, Pid: pID, Signal: gnps.KillProcessRequest_SIGNAL_KILL, Restart: true}
+	killResponse, err := gnoiClient.System().KillProcess(context.Background(), killRequest)
+	t.Logf("Got kill process response: %v\n\n", killResponse)
+	if err != nil {
+		t.Fatalf("Failed to execute gNOI Kill Process, error received: %v", err)
+	}
+}
+
+func TestDUTDaemonFailure(t *testing.T) {
+
+	start := time.Now()
+	dut := ondatra.DUT(t, "dut")
+	ctx := context.Background()
+
+	// Configure the DUT.
+	t.Logf("Configure DUT")
+	configureDUT(t, dut)
+
+	// Configure the ATE.
+	t.Logf("Configure ATE")
+	ate := ondatra.ATE(t, "ate")
+	top := configureATE(t, ate)
+	top.Push(t).StartProtocols(t)
+
+	t.Logf("Time check: %s", time.Since(start))
+
+	args := &testArgs{
+		ctx: ctx,
+		dut: dut,
+		ate: ate,
+		top: top,
+	}
+
+	t.Run("SetupGribiConnection", func(t *testing.T) {
+
+		// Set parameters for gRIBI client clientA.
+		// Set Persistence to true.
+		clientA := &gribi.Client{
+			DUT:                  dut,
+			FibACK:               false,
+			Persistence:          true,
+			InitialElectionIDLow: 10,
+		}
+
+		t.Log("Establish gRIBI client connection")
+		if err := clientA.Start(t); err != nil {
+			t.Fatalf("gRIBI Connection for clientA could not be established")
+		}
+
+		t.Run("AddRoute", func(t *testing.T) {
+			t.Logf("Add gRIBI route to %s and verify through Telemetry and Traffic", ateDstNetCIDR)
+			addRoute(ctx, t, args, clientA)
+
+			t.Run("VerifyAFT", func(t *testing.T) {
+				verifyAFT(ctx, t, args)
+			})
+
+			t.Run("VerifyTraffic", func(t *testing.T) {
+				verifyTraffic(ctx, t, args)
+			})
+
+		})
+
+		t.Logf("Time check: %s", time.Since(start))
+
+	})
+
+	t.Run("KillGribiDaemon", func(t *testing.T) {
+
+		// Find the PID of gRIBI Daemon.
+		var pName string = "unknown"
+		var pId uint64
+		t.Run("FindGribiDaemonPid", func(t *testing.T) {
+			switch dut.Vendor() {
+			case ondatra.JUNIPER:
+				pName = "rpd"
+			case ondatra.CISCO:
+				pName = "unknown-cisco" //TODO
```

Comment:
@mingyangcisco what is the process name of the gRIBI daemon on Cisco? This is for TE-8.1 to kill gRIBI process.

Reply:
@liulk it's `emsd`

Reply:
@liulk please note that the emsd process is a common host process for all grpc services (e.g. gnoi, gnmi and gribi). 

This test seems to not take this architecture into account, and I believe will fail on Cisco until the process restarts. The way this test is written it expects gnmi to respond, which will not work for the Cisco architecture. I think maybe a reevaluation of the test case may be warranted.

Reply:
I don't know if it's ever the intention for test plans to define the de-facto behavior of the implementation, but I think the only way gNOI.KillProcess can be implemented sensibly when the process being killed happens to be the daemon handing the KillProcess request, is if the daemon will first return success, then kill itself after a short delay.

If the daemon kills itself without cleanly returning, the client will get something like "socket closed (unavailable 14)" as the error. As it turns out, our internal binding will implicitly retry on socket closed, which means it will repeatedly send KillRequest. It would be funny to watch, but I can't think of any way to avoid this retry.

But having said this, I'm not the one who designed this test. Perhaps @dplore can point you to the right person.

Reply:
@aaronmillisor I think delaying the streaming telemetry validation should be fine as it does not change the focus/goal of this test (as you said there is still traffic flowing to assist validation). 

@aaronmillisor does `emsd` also handle gNOI request? If so, maybe your gNOI implementation might want to consider some of @liulk 's point in terms of "killing the daemon itself"? 

Thanks.

Reply:
if we use `KillProcessRequest_SIGNAL_TERM` signal,  the restart happens after replying the GNOI request. But using `KillProcessRequest_SIGNAL_KILL` signal will terminate the connection right away and will result in GNOI request failure.